kaspresso/src/main/kotlin/com/kaspersky/kaspresso/autoscroll/AutoScrollFallbackProviderImpl.kt

```diff
@@ -0,0 +1,129 @@
+package com.kaspersky.kaspresso.autoscroll
+
+import android.util.Log
+import android.view.View
+import android.widget.HorizontalScrollView
+import android.widget.ListView
+import android.widget.ScrollView
+import androidx.core.widget.NestedScrollView
+import androidx.test.espresso.PerformException
+import androidx.test.espresso.UiController
+import androidx.test.espresso.ViewAction
+import androidx.test.espresso.ViewInteraction
+import androidx.test.espresso.matcher.ViewMatchers
+import androidx.test.espresso.util.HumanReadables
+import com.kaspersky.kaspresso.internal.extensions.other.isAllowed
+import com.kaspersky.kaspresso.logger.UiTestLogger
+import com.kaspersky.kaspresso.params.AutoScrollParams
+import org.hamcrest.Matcher
+import org.hamcrest.Matchers
+import androidx.test.espresso.matcher.ViewMatchers.isDisplayingAtLeast
+
+
+class AutoScrollFallbackProviderImpl(
+    private val params: AutoScrollParams,
+    private val logger: UiTestLogger
+) : AutoScrollProvider<ViewInteraction> {
+
+    /**
+     * Invokes the given [action] and calls [scroll] if it fails. Helps in cases when test fails because of the
+     * need to scroll to interacted view.
+     *
+     * @param interaction the instance of [ViewInteraction] interface to perform actions and assertions.
+     * @param action the actual action on the interacted view.
+     *
+     * @throws Throwable if the exception caught while invoking [action] is not allowed via [params].
+     * @return the result of [action] invocation.
+     */
+    @Throws(Throwable::class)
+    override fun <T> withAutoScroll(interaction: ViewInteraction, action: () -> T): T {
+        return try {
+            action.invoke()
+        } catch (error: Throwable) {
+            if (error.isAllowed(params.allowedExceptions)) {
+                return scroll(interaction, action, error)
+            }
+            throw error
+        }
+    }
+
+    /**
+     * Performs the autoscrolling functionality. Performs scroll and re-invokes the given [action].
+     *
+     * @param interaction the instance of [ViewInteraction] interface to perform actions and assertions.
+     * @param action the actual action on the interacted view.
+     * @param cachedError the error to be thrown if autoscroll would not help.
+     *
+     * @throws cachedError if autoscroll action did not help.
+     * @return the result of [action] invocation.
+     */
+    @Throws(Throwable::class)
+    override fun <T> scroll(interaction: ViewInteraction, action: () -> T, cachedError: Throwable): T {
+        return try {
+            interaction.perform(FallbackScrollToAction())
+            logger.i("View autoScroll fallback successfully performed.")
+            action.invoke()
+        } catch (error: Throwable) {
+            throw cachedError
+        }
+    }
+}
+
+class FallbackScrollToAction : ViewAction {
```

Comment:
I nested the action here. This is the actual solution.
It makes sense to move this to another folder.

I am actually thinking that this could be added to Kakao library indeed?

What are your thoughts on this?
@Vacxe @matzuk

Reply:
@Vacxe ping